### PR TITLE
[Feature] Add Cylinder3D Backbone

### PR DESCRIPTION
**Pull request by xizaoqu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add Cylinder3D Backbone.

## Modification

Add Cylinder3D Backbone.

## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet3d/models/backbones/cylinder3d.py [hunk 1]
```diff
@@ -0,0 +1,465 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+r"""Modified from Cylinder3D.
+
+Please refer to `Cylinder3D github page <http://www.semantic-
```
Comment (Tai-Wang) [inline mmdet3d/models/backbones/cylinder3d.py:4]: Please revise the link

File: mmdet3d/models/backbones/cylinder3d.py [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+        in_planes,
+        out_planes,
+        kernel_size=1,
+        stride=stride,
+        padding=1,
+        bias=False,
+        indice_key=indice_key)
+
+
+class AsymmResBlock(nn.Module):
```
Comment (Tai-Wang) [inline mmdet3d/models/backbones/cylinder3d.py:96]: May ablate the inheritation from nn.Module and BaseModule

Comment (Tai-Wang) [reply mmdet3d/models/backbones/cylinder3d.py:96]: And please add typehint for these classes

File: mmdet3d/models/backbones/cylinder3d.py [hunk 3]
```diff
@@ -0,0 +1,576 @@
...
+        out_channels (int): Output channels of the block.
+        norm_cfg (dict or ConfigDict): Config to construct and config
+            norm layer.
+        indice_key (str, optional): Name of indice tables. Default: None.
+    """
+
+    def __init__(self,
+                 in_channels: int,
+                 out_channels: int,
+                 norm_cfg: dict,
```
Comment (ZCMax) [inline mmdet3d/models/backbones/cylinder3d.py:208]: check the type hint

File: mmdet3d/models/backbones/cylinder3d.py [hunk 4]
```diff
@@ -0,0 +1,576 @@
...
+           block. Defaults: True.
+       height_pooling (bool): Whether pooling features at
+           the height dimension. Defaults: False.
+       indice_key (str, optional): Name of indice tables. Default: None.
+    """
+
+    def __init__(self,
+                 in_channels: int,
+                 out_channels: int,
+                 norm_cfg: dict,
```
Comment (ZCMax) [inline mmdet3d/models/backbones/cylinder3d.py:274]: same here

File: mmdet3d/models/backbones/cylinder3d.py [hunk 5]
```diff
@@ -0,0 +1,576 @@
...
+        up.features = self.bn2(up.features)
+
+        up = self.conv3(up)
+        up.features = self.act3(up.features)
+        up.features = self.bn3(up.features)
+
+        return up
+
+
+class DDCMBlock(nn.Module):
```
Comment (ZCMax) [inline mmdet3d/models/backbones/cylinder3d.py:423]: BaseModule

File: mmdet3d/models/backbones/cylinder3d.py [hunk 6]
```diff
@@ -0,0 +1,578 @@
...
+        return up
+
+
+class DDCMBlock(BaseModule):
+    """Dimension-Decomposition based Context Modeling.
+
+    Args:
+        in_channels (int): Input channels of the block.
+        out_channels (int): Output channels of the block.
+        norm_cfg (dict pr Config): Config to construct and config norm layer.
```
Comment (ZCMax) [inline mmdet3d/models/backbones/cylinder3d.py:430]: pr -> or

File: mmdet3d/models/backbones/cylinder3d.py [hunk 7]
```diff
@@ -0,0 +1,578 @@
...
+    Args:
+       in_channels (int): Input channels of the block.
+       out_channels (int): Output channels of the block.
+       norm_cfg (:obj:`ConfigDict` or dict): Config dict for
+            normalization layer.
+       pooling (bool): Whether pooling features at the end of
+           block. Defaults: True.
+       height_pooling (bool): Whether pooling features at
+           the height dimension. Defaults: False.
+       indice_key (str, optional): Name of indice tables. Default: None.
```
Comment (ZCMax) [inline mmdet3d/models/backbones/cylinder3d.py:269]: Change all `Default: xxx` to `Defaults to ...`

File: mmdet3d/models/backbones/cylinder3d.py [hunk 8]
```diff
@@ -0,0 +1,581 @@
...
+from torch import nn as nn
+
+from mmdet3d.registry import MODELS
+from mmdet3d.utils import OptConfigType
+
+
+def conv3x3x3(in_channels: int,
+              out_channels: int,
+              stride: int = 1,
+              indice_key: str = None) -> SparseConvTensor:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/cylinder3d.py:25]: SparseConvTensor -> SparseModule. Please also check other typehint

Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/cylinder3d.py:25]: Optional[str]

File: mmdet3d/models/backbones/cylinder3d.py [hunk 9]
```diff
@@ -0,0 +1,581 @@
...
+        out_channels (int): Output channels of the block.
+        norm_cfg (Union[:obj:`ConfigDict` or dict]): Config dict for
+            normalization layer.
+        indice_key (str, optional): Name of indice tables. Defaults to None.
+    """
+
+    def __init__(self,
+                 in_channels: int,
+                 out_channels: int,
+                 norm_cfg: Union[OptConfigType, dict],
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/cylinder3d.py:211]:
```suggestion
                 norm_cfg: ConfigType,
```

File: mmdet3d/models/backbones/cylinder3d.py [hunk 10]
```diff
@@ -0,0 +1,581 @@
...
+           block. Defaults: True.
+       height_pooling (bool): Whether pooling features at
+           the height dimension. Defaults: False.
+       indice_key (str, optional): Name of indice tables. Defaults to None.
+    """
+
+    def __init__(self,
+                 in_channels: int,
+                 out_channels: int,
+                 norm_cfg: Union[OptConfigType, dict],
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/cylinder3d.py:277]: The same as above.

File: mmdet3d/models/backbones/cylinder3d.py [hunk 11]
```diff
@@ -0,0 +1,580 @@
...
+            up_key='down1',
+            norm_cfg=norm_cfg)
+        self.up_block3 = AsymmeUpBlock(
+            4 * init_size,
+            2 * init_size,
+            indice_key='up3',
+            up_key='down0',
+            norm_cfg=norm_cfg)
+
+        self.ddcm = DDCMBlock(
```
Comment (ZwwWayne) [inline mmdet3d/models/backbones/cylinder3d.py:556]: cannot guess what ddcm mean from the abbreviation, try more clear name

File: mmdet3d/models/backbones/cylinder3d.py [hunk 12]
```diff
@@ -0,0 +1,580 @@
...
+       indice_key (str, optional): Name of indice tables. Defaults to None.
+       up_key (str, optional): Name of indice tables used in
+           SparseInverseConv3d. Defaults to None.
+    """
+
+    def __init__(self,
+                 in_channels: int,
+                 out_channels: int,
+                 norm_cfg: ConfigType,
+                 indice_key: Optional[str] = None,
```
Comment (ZwwWayne) [inline mmdet3d/models/backbones/cylinder3d.py:368]: conside act_cfg and use build_activation_layer(act_cfg)

File: mmdet3d/models/backbones/cylinder3d.py [hunk 13]
```diff
@@ -0,0 +1,580 @@
...
+        out_channels (int): Output channels of the block.
+        norm_cfg (:obj:`ConfigDict` or dict): Config dict for
+            normalization layer.
+        indice_key (str, optional): Name of indice tables. Defaults to None.
+    """
+
+    def __init__(self,
+                 in_channels: int,
+                 out_channels: int,
+                 norm_cfg: ConfigType,
```
Comment (ZwwWayne) [inline mmdet3d/models/backbones/cylinder3d.py:439]: add act_cfg and use build_activation_layer(act_cfg)

File: mmdet3d/models/backbones/cylinder3d.py [hunk 14]
```diff
@@ -0,0 +1,580 @@
...
+            in_channels, out_channels, indice_key=indice_key + 'bef')
+        self.act1_0 = nn.LeakyReLU()
+        self.bn1_0 = build_norm_layer(norm_cfg, out_channels)[1]
+
+        self.conv1_1 = conv1x3x3(
+            out_channels, out_channels, indice_key=indice_key + 'bef')
+        self.act1_1 = nn.LeakyReLU()
+        self.bn1_1 = build_norm_layer(norm_cfg, out_channels)[1]
+
+    def forward(self, x: SparseModule) -> SparseModule:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/cylinder3d.py:235]: SparseModule -> SparseConvTensor.

File: mmdet3d/models/backbones/cylinder3d.py [hunk 15]
```diff
@@ -0,0 +1,580 @@
...
+                self.pool = SparseConv3d(
+                    out_channels,
+                    out_channels,
+                    kernel_size=3,
+                    stride=(2, 2, 1),
+                    padding=1,
+                    indice_key=indice_key,
+                    bias=False)
+
+    def forward(self, x: SparseModule) -> SparseModule:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/cylinder3d.py:324]: SparseModule -> SparseConvTensor

File: mmdet3d/models/backbones/cylinder3d.py [hunk 16]
```diff
@@ -0,0 +1,580 @@
...
+        self.bn3 = build_norm_layer(norm_cfg, out_channels)[1]
+
+        self.up_subm = SparseInverseConv3d(
+            out_channels,
+            out_channels,
+            kernel_size=3,
+            indice_key=up_key,
+            bias=False)
+
+    def forward(self, x: SparseModule, skip: SparseModule) -> SparseModule:
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/cylinder3d.py:399]: SparseModule -> SparseConvTensor, and check others.

File: mmdet3d/models/backbones/cylinder3d.py (hunk 11 above)
Comment (xizaoqu) [reply mmdet3d/models/backbones/cylinder3d.py:556]: I mentioned the meaning in the docstrings of DDCMBlock.

File: mmdet3d/models/backbones/cylinder3d.py [hunk 17]
```diff
@@ -0,0 +1,592 @@
...
+        init_size (int): Initial size of feature channels before
+            feeding into Encoder-Decoder structure. Defaults to 16.
+        norm_cfg (:obj:`ConfigDict` or dict): Config dict for normalization
+            layer. Defaults to dict(type='BN1d', eps=1e-3, momentum=0.01)).
+    """
+
+    def __init__(self,
+                 grid_size: int,
+                 input_dims: int,
+                 init_size: int = 16,
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/cylinder3d.py:507]: init_size -> base_channels

File: mmdet3d/models/backbones/cylinder3d.py [hunk 18]
```diff
@@ -0,0 +1,592 @@
...
+        input_dims (int): Input channels of the block.
+        init_size (int): Initial size of feature channels before
+            feeding into Encoder-Decoder structure. Defaults to 16.
+        norm_cfg (:obj:`ConfigDict` or dict): Config dict for normalization
+            layer. Defaults to dict(type='BN1d', eps=1e-3, momentum=0.01)).
+    """
+
+    def __init__(self,
+                 grid_size: int,
+                 input_dims: int,
```
Comment (Xiangxu-0103) [inline mmdet3d/models/backbones/cylinder3d.py:506]: input_dims -> in_channels

**Comment by xizaoqu** [comment]
fix CLA error

**Pull request closed without merging**